dandi/download.py

```diff
@@ -501,3 +502,68 @@ def _download_file(
         os.utime(path, (time.time(), ensure_datetime(mtime).timestamp()))
 
     yield {"status": "done"}
+
+
+class DownloadDirectory:
+    def __init__(self, filepath, filehash):
+        #: The path to which to save the file after downloading
+        self.filepath = Path(filepath)
+        #: The expected hash of the downloaded data
+        self.filehash = filehash
+        #: The working directory in which downloaded data will be temporarily
+        #: stored
+        self.dirpath = append_ext(self.filepath, ".download")
+        #: The file in `dirpath` to which data will be written as it is
+        #: received
+        self.writefile = self.dirpath / self.filepath.name
+        #: A `fasteners.InterProcessLock` on `dirpath`
+        self.lock = None
+        #: An open filehandle to `writefile`
+        self.fp = None
+        #: How much of the data has been downloaded so far
+        self.offset = None
+
+    def __enter__(self):
+        from fasteners import InterProcessLock
+
+        self.dirpath.mkdir(parents=True, exist_ok=True)
+        self.lock = InterProcessLock(str(append_ext(self.writefile, ".lock")))
+        if not self.lock.acquire(blocking=False):
+            raise RuntimeError("Could not acquire lock on download")
+        chkpath = append_ext(self.writefile, ".checksum")
+        try:
+            checksum = chkpath.read_text(encoding="utf-8").strip()
+        except FileNotFoundError:
+            checksum = ""
+        if checksum == self.filehash:
+            # Pick up where we left off, writing to the end of the file
+            self.fp = self.writefile.open("ab")
+        else:
+            # Truncate the file (if it even exists)
+            self.fp = self.writefile.open("wb")
+            chkpath.write_text(str(self.filehash), encoding="utf-8")
+        self.offset = self.fp.tell()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if exc_type is None:
+            from shutil import rmtree
+
+            self.fp.close()
```

Comment:
`self.fp.close()` common to both cases - so why not to have it before `if`? and shouldn't `self.lock.release()` go at the end  of the handling (after `if`).  Paranoid me would also wrap the rest (renaming, rmtree) within `try/except` so if somehow those fail (permissions!?) we still release the lock and then reraise.

Reply:
I've adjusted the method.  There's no need to wrap `rmtree()` in a `try`, as it has `ignore_errors=True`.